Spells able to cost both mana and manapercent at once and also health and healthpercent +luascript

Olrios opened the pull request:
I made spells able to cost for example 20 mana and 5% max mana at once.
Would be useful for custom servers. If this code is okey and the way I published it tell me please.

Try it out with:

In spells.xml you can add this code to any spell:

`soul="x" mana="x" manapercent="x" health="x" healthpercent="x"`
soul was before; mana and manapercent was before too, but they didn't worked together. Now you can use them all in any combination.

You can add for example `health="40"` to your spell in spells.xml and everytime you cast the spell, it will cost you 40 health. Also your magic level will grow by 40 ticks. If your health is below 40, then you wont be able to cast the spell.

also you can use this in lua, so you can also update your spellsbook.lua:

```
for _, spell in ipairs(player:getInstantSpells()) do
...
spell.mana -- this was before
spell.manapercent -- this was before
spell.health
spell.healthpercent
spell.soul
```

Add `healthCostSpellsTrainMagic = true` to your config.lua to make you able to train your mlevel with health-cost spells.

Fixes #2733

Olrios commented:
All fixed. Tested it with multiple setups - works perfect.
You can even set soul="1" mana="x" manapercent="x" health="x" healthpercent="x" at once in spells.xml

Znote commented:
Tagging related issue. 

Fix #2733

Olrios commented:
In spells.xml you can add this code to any spell:

`soul="x" mana="x" manapercent="x" health="x" healthpercent="x"`

also you can use in lua this:

```
spell = player:getInstantSpells()[1]
...
spell.mana
spell.manapercent
spell.health
spell.healthpercent
spell.soul
```

EvilHero90 commented:
> 
> 
> Seems to work great when registering with XML, but does not work properly in Revscriptsys:
> 
> data/**scripts**/PR_quicktest.lua
> 
> ```lua
> local combat = Combat()
> combat:setParameter(COMBAT_PARAM_TYPE, COMBAT_FIREDAMAGE)
> combat:setParameter(COMBAT_PARAM_EFFECT, CONST_ME_FIREAREA)
> combat:setArea(createCombatArea(AREA_CIRCLE5X5))
> 
> function onGetFormulaValues(player, level, magicLevel)
>    local min = (level / 4) + (magicLevel * 8) + 50
>    local max = (level / 4) + (magicLevel * 16) + 75
>    return -min, -max
> end
> 
> combat:setCallback(CALLBACK_PARAM_LEVELMAGICVALUE, "onGetFormulaValues")
> 
> local spell = Spell("instant")
> 
> function spell.onCastSpell(creature, variant)
>    return combat:execute(creature, variant)
> end
> 
> spell:name("Fire Storm")
> spell:words("exevo flam inferno")
> spell:group("attack")
> spell:vocation("sorcerer", "master sorcerer")
> spell:cooldown(4 * 1000)
> spell:groupCooldown(2 * 1000)
> spell:level(60)
> spell:mana(30)
> spell:healthpercent(100)
> spell:soul(2)
> spell:isSelfTarget(true)
> spell:isPremium(false)
> 
> spell:register()
> ```
> 
> Console:
> 
> ```
> Lua Script Error: [Scripts Interface]
> /mnt/c/Users/Stefan/Documents/ubuntu18/znote/data/scripts/PR_quicktest.lua
> ...n/Documents/ubuntu18/znote/data/scripts/PR_quicktest.lua:28: attempt to call method 'healthpercent' (a nil value)
> stack traceback:
>         [C]: in function 'healthpercent'
> ```
> 
> Maybe @EvilHero90 has some tips?

```healthpercent``` has to be ```healthPercent```

Znote commented:
Thanks, that worked. 

Next problem: 
![image](https://user-images.githubusercontent.com/1552144/76685110-71ec2500-6611-11ea-81e5-2731ccdf09d3.png)

I have tons of mana. What I'm lacking is health.

EvilHero90 commented on src/spells.cpp:
```diff
@@ -637,6 +645,12 @@ bool Spell::playerSpellCheck(Player* player) const
 	}
 
+	if (player->getHealth() < static_cast<int32_t>(getHealthCost(player)) && !player->hasFlag(PlayerFlag_HasInfiniteMana)) {
+		player->sendCancelMessage(RETURNVALUE_NOTENOUGHMANA);
```
needs a new return value, so it shows the correct message

Olrios replied:
fixed, added 'not enought hp' msg

EvilHero90 commented on src/spells.cpp:
```diff
@@ -819,17 +833,22 @@ void Spell::postCastSpell(Player* player, bool finishedCast /*= true*/, bool pay
 	}
 
+	if (healthCost > 0) {
+		player->addManaSpent(healthCost);
```
it shouldn't add mana spent as this could be easily abused, mana is mana and health is health this should stay separated

Olrios replied:
without this using spells what cost health only will not train your magic level at all

Olrios replied:
maybe ill add an option for it in config.lua

Olrios replied:
fixed, added option for it in config.lua

EvilHero90 commented on src/luascript.cpp:
```diff
@@ -14204,6 +14228,25 @@ int LuaScriptInterface::luaSpellManaPercent(lua_State* L)
+			pushBoolean(L, true);
+		}
+	}
+	else {
```
indentation ```} else {```

Olrios replied:
fixed

EvilHero90 commented on src/luascript.cpp:
```diff
@@ -14204,6 +14228,25 @@ int LuaScriptInterface::luaSpellManaPercent(lua_State* L)
+		if (lua_gettop(L) == 1) {
+			lua_pushnumber(L, spell->getHealthPercent());
+		}
+		else {
```
indentation ```} else {```

Olrios replied:
fixed

EvilHero90 commented on src/luascript.cpp:
```diff
@@ -14187,6 +14192,25 @@ int LuaScriptInterface::luaSpellMana(lua_State* L)
+			pushBoolean(L, true);
+		}
+	}
+	else {
```
indentation ```} else {```

Olrios replied:
fixed

EvilHero90 commented on src/luascript.cpp:
```diff
@@ -14187,6 +14192,25 @@ int LuaScriptInterface::luaSpellMana(lua_State* L)
+		if (lua_gettop(L) == 1) {
+			lua_pushnumber(L, spell->getHealth());
+		}
+		else {
```
indentation ```} else {```

Olrios replied:
fixed

Olrios commented:
My spellbook.lua:
`function onUse(player, item, fromPosition, target, toPosition, isHotkey)
	local text = ""
	local costText, manaCost, healthCost, soulCost = "", "", "", ""
	local spells = {}
	for _, spell in ipairs(player:getInstantSpells()) do
		if spell.level ~= 0 then
			local costText, manaCost, healthCost, soulCost = "", "", "", ""
		
			if spell.mana > 0 then
				manaCost = spell.mana
			end
			if spell.mana > 0 and spell.manapercent > 0 then
				manaCost = manaCost.. " + "
			end
			if spell.manapercent > 0 then
				manaCost = manaCost..spell.manapercent.. "%"
			end
			if manaCost ~= "" then
				manaCost = manaCost.." mana"
				costText = manaCost
			end
			
			if spell.health > 0 then
				healthCost = spell.health
			end
			if spell.health > 0 and spell.healthpercent > 0 then
				healthCost = healthCost.. " + "
			end
			if spell.healthpercent > 0 then
				healthCost = healthCost..spell.healthpercent.. "%"
			end
			if healthCost ~= "" then
				healthCost = healthCost.." health"
				if manaCost ~= "" then
					costText = costText.. " and "
				end
				costText = costText..healthCost
			end
			
			if spell.soul > 0 then
				soulCost = spell.soul.. " soul"
				if spell.soul > 1 then
					soulCost = soulCost.."s"
				end
				if costText ~= "" then
					costText = costText.. " and "
				end
				costText = costText..soulCost
			end
			
			spells[#spells + 1] = spell
		end
	end

	table.sort(spells, function(a, b) return a.level < b.level end)

	local prevLevel = -1
	for i, spell in ipairs(spells) do
		local line = ""
		if prevLevel ~= spell.level then
			if i ~= 1 then
				line = "\n"
			end
			line = line .. "Spells for Level " .. spell.level .. "\n"
			prevLevel = spell.level
		end
		text = text .. line .. "  " .. spell.name .. " - " .. costText .. "\n"
	end

	player:showTextDialog(item:getId(), text)
	return true
end
`

Works without any bugs.
Also I didn't implement "don't have enought health" message. It uses instead "don't have enought mana" right now.
Also without this part: `player->addManaSpent(healthCost);` using spells what costs health only will not train your magic level at all.

Olrios commented:
1. Fixed all `} else {`.
2. Now you can use `healthCostSpellsTrainMagic = true` in config.lua to make health-cost spells able to train your magic level.
3. Added NOT_ENOUGH_HEALTH.

Olrios commented:
Added fix. Had to do a little workaround. This how code handles GM's mana and infinite_mana_flag or infinite_soul is a little weird and should be rewritten. Saw some infinite_soul inside `void Spell::postCastSpell` but infinite mana flag is handled in different way.

DSpeichert commented:
Isn't the spell book supposed to be updated as well?

EvilHero90 commented on src/spells.cpp:
```diff
@@ -637,6 +645,12 @@ bool Spell::playerSpellCheck(Player* player) const
 		return false;
 	}
 
+	if (player->getHealth() < static_cast<int32_t>(getHealthCost(player)) && !player->hasFlag(PlayerFlag_HasInfiniteMana)) {
```
should be
```c++
if (player->getHealth() <= ...
```
because the player could die if the ammount is exactly the same as his current health is.

Olrios replied:
How i didn't notice this? Thanks for finding it. Fixed.

nekiro commented on src/spells.cpp:
```diff
@@ -637,6 +645,12 @@ bool Spell::playerSpellCheck(Player* player) const
 		return false;
 	}
 
+	if (player->getHealth() <= static_cast<int32_t>(getHealthCost(player)) && !player->hasFlag(PlayerFlag_HasInfiniteMana)) {
```
PlayerFlag_HasInfiniteMana? Shouldn't this be new flag HasInfiniteHealth?

nekiro commented on src/spells.cpp:
```diff
@@ -819,17 +833,26 @@ void Spell::postCastSpell(Player* player, bool finishedCast /*= true*/, bool pay
+		if (g_config.getBoolean(ConfigManager::HEALTH_COST_SPELLS_TRAIN_MAGIC)) {
+			player->addManaSpent(healthCost);
+		}
+		if (player->getAccountType() < ACCOUNT_TYPE_GAMEMASTER && !player->isAccessPlayer()) {
```
PlayerFlag_NotGainHealth

nekiro commented on src/spells.cpp:
```diff
@@ -819,17 +833,26 @@ void Spell::postCastSpell(Player* player, bool finishedCast /*= true*/, bool pay
 
+	if (healthCost > 0) {
+		if (g_config.getBoolean(ConfigManager::HEALTH_COST_SPELLS_TRAIN_MAGIC)) {
+			player->addManaSpent(healthCost);
```
additionaly you should check the flag PlayerFlag_NotGainSkill
